rs/replicated_state/src/canister_state/system_state.rs

```diff
@@ -1405,10 +1408,22 @@ impl SystemState {
             if own_subnet_type != SubnetType::System
                 && can_push(msg, *subnet_available_memory).is_err()
             {
-                // Bail out if not enough memory available for message.
                 return;
             }
 
+            // Protect against enqueuing a second response for the currently executing
+            // (aborted or paused) callback.
+            if let RequestOrResponse::Response(response) = &msg {
+                if let Some(aboirted_or_paused_response) = self.aborted_or_paused_response() {
+                    if response.originator_reply_callback
+                        == aboirted_or_paused_response.originator_reply_callback
+                    {
+                        // The callback is already executing, don't enqueue a second response for it.
+                        return;
```

Comment:
Is returning here the right thing to do? I suppose returning means that things get handled via the loopback stream leading to eventually dropping the duplicate message upon induction?

Reply:
Returning stops inducting messages. We peeked a message we didn't like and returned without popping it or the messages after it.